pkg/envoy/ads/grpc.go

```diff
@@ -28,26 +30,45 @@ func receive(requests chan xds_discovery.DiscoveryRequest, server *xds_discovery
 		}
 		if !proxy.HasPodMetadata() {
 			// Set the Pod metadata on the given proxy only once. This could arrive with the first few XDS requests.
-			recordEnvoyPodMetadata(request, proxy, proxyRegistry)
+			if err := recordEnvoyPodMetadata(request, proxy, proxyRegistry); err != nil {
+				log.Err(err).Msgf("Error recording Pod metadata")
+				// this terminates the gRPC stream
+				return
+			}
 		}
 		log.Trace().Msgf("[grpc] Received DiscoveryRequest from Envoy with certificate SerialNumber %s", proxy.GetCertificateSerialNumber())
 		requests <- *request
 	}
 }
 
-func recordEnvoyPodMetadata(request *xds_discovery.DiscoveryRequest, proxy *envoy.Proxy, proxyRegistry *registry.ProxyRegistry) {
+func recordEnvoyPodMetadata(request *xds_discovery.DiscoveryRequest, proxy *envoy.Proxy, proxyRegistry *registry.ProxyRegistry) error {
```

Comment:
I'd add a test for this function and the newly added code.

Reply:
+1

Reply:
Made https://github.com/openservicemesh/osm/issues/3363 for this - to follow up shortly.